### PR TITLE
Get the sha1 from the file directly, not through its public URI

### PR DESCRIPTION
**Pull request by baschny**
1. We should not try to fetch an "Public URI" to apply SHA1 to it, as it won't work:

a) if the URI is not reachable (i.e. in Monocle)
b) if PHP does not allow calling "external URIs" in its stream functions (i.e. with `open_basedir` restriction)

This should make it more robust, but also make it only work with "resources" from our own packages (which is what we intend anyway).

Also fix the URI if it already might contain a query parameter with "?" to use a "&" separator instead.

2. also add the "path" as a cache tag, which allows us to clear a particular cached entry if required.

### CODE REVIEW

**Review by marco-wildermuth — APPROVED** [review]

File: Classes/FusionObjects/HashedResourceUriImplementation.php [hunk 1]
```diff
@@ -2,23 +2,58 @@
...
+                throw new Exception(sprintf('The specified path "%s" does not point to a valid package.', $path), 1617009246);
+            }
+            $filename = $packagePath . FlowPackageInterface::DIRECTORY_RESOURCES . $relativePath;
+            if (!is_readable($filename)) {
+                throw new Exception(sprintf('The specified path "%s" does not point to a readable filename.', $path), 1617009245);
+            }
 
-        return $resourceUri . ($sha1 ? '?' . $sha1 : '');
+            $sha1 = sha1_file($filename);
+            $resourceUri .= (strpos($resourceUri, '?') === false) ? '?' : '&';
```
Comment (remuslazar) [inline Classes/FusionObjects/HashedResourceUriImplementation.php:52]: @baschny instead of doing the "hard work" manually, the PHP' `parse_url` function could be used here. Then you could use the `scheme` and `query` and be more high level:

https://www.php.net/manual/en/function.parse-url.php

Comment (baschny) [reply Classes/FusionObjects/HashedResourceUriImplementation.php:52]: Interesting. But how would you "put it back together"? I don't know of an API for that. The parse_url returns just an array of "strings". Or is there are reverse method to do that without having to manually concatenate the strings?

Comment (baschny) [reply Classes/FusionObjects/HashedResourceUriImplementation.php:52]: Just found out `http_build_url`. Never head of that. Next time, will try it. :)

Comment (remuslazar) [reply Classes/FusionObjects/HashedResourceUriImplementation.php:52]: PHP has some good stuff built in, beside of the thousand `array_*` functions :)

TL;DR

But I thought to skip the regex magic you are doing just for parsing. Appending the parameter, using ? or & could then be done just by inspecting the `query` value from `parse_url` retval then.

And not having to search for the ? using `strpos` which I personally find a very weird pattern and not very expressive :)
